core/src/main/scala/scalaz/Foldable.scala

```diff
@@ -364,16 +365,12 @@ trait Foldable[F[_]]  { self =>
    * Selects groups of elements that satisfy p and discards others.
    */
   def selectSplit[A](fa: F[A])(p: A => Boolean): List[NonEmptyList[A]] =
-    foldRight(fa, (List[NonEmptyList[A]](), false))((a, xb) => xb match {
-      case (x, b) => {
-        val pa = p(a)
-        (if (pa)
-          if (b)
-            (a <:: x.head) :: x.tail else
-            NonEmptyList(a) :: x
-        else x, pa)
-      }
-    })._1
+    foldRight(fa, List.empty[NonEmptyList[A]])((a, li) => (li, p(a)) match {
+      case (x :: xs, pa) =>
+        if (pa) (a <:: x) :: xs else li
+      case (Nil, pa) =>
+        if (pa) NonEmptyList(a) :: li else li
```

Comment:
Doesn't this implementation return at most one `NonEmptyList`?

Reply:
Let's include some test to protect against it.

Reply:
If initial `F[_]` is empty - the result will be empty as well.